### PR TITLE
AKS Service Target - Overriding KUBECONFIG handling

### PR DESCRIPTION
**Pull request by wbreza**
This change builds on smaller set of changes from other open PRs:

- #2455 
- #2464

Additionally this adds support to override behavior of the `KUBECONFIG` environment variable.

## Use Case: Standard `azd` use case

When provisioning is complete `azd` will automatically attempt to authenticate to the cluster with local admin credentials.  A new `KUBECONFIG` is created with the same name as the AKS cluster name and merged into the default `KUBECONFIG`.  

The default context within the `KUBECONFIG` now also includes the default project/service namespace.

> **Note**
> Namespaces can be set within resource yaml definitions.  If not specified `azd` will use the default service namespace. Services can specify namespace name in the `k8s` section of the service yaml.  
> 
> If no namespace has been set for the service the `azd` project name will be used by default.

The k8s current context is set to the new cluster with default project namespace.  At this point any custom `kubectl` commands that exist within `postprovision` or `predeploy` scripts can make `kubectl` commands and it will automatically use the new `KUBECONFIG`.

## Use Case: Overriding `KUBECONFIG` environment variable

If `KUBECONFIG` is set in either the azd `.env` file or in the OS `azd` will use this `KUBECONFIG` as its default.  `azd` will **_NOT_** create a new `KUBECONFIG` or context and the developer must ensure the `KUBECONFIG` is initialized with correct authentication credentials with context and namespace.

Any calls to `kubectl` within `postprovision` or `predeploy` scripts will automatically use the specified `KUBECONFIG`.

### CODE REVIEW

**Comment by carmodyquinnms** [comment]
Tested this with azd and there seems to be an issue with the permission level on the kubeconfig that is being pulled.  I have local accounts enabled on AKS however this is pulling credentials where it hangs awaiting a browser to be opened and authorized.  But it seems to be hanging in the postprovision phase as opposed to predeployment?

**Comment by cquinn-dcsg** [comment]
@wbreza It seems as though the templating feature is no longer working on the PR.   When deploying the code azd fails on https://github.com/Azure-Samples/ratings-nodejs-azd-akslza/blob/cq-ghworkflow/src/manifests/api/1-cert-manager.yaml
with errors on the namespace.  It was allowing any manifest file without  *.tmpl.yaml to be applied directly but that has stopped working.  Here are the errors from "azd deploy --debug"

clusterrolebinding.rbac.authorization.k8s.io/cert-manager-webhook:subjectaccessreviews created
mutatingwebhookconfiguration.admissionregistration.k8s.io/cert-manager-webhook created
validatingwebhookconfiguration.admissionregistration.k8s.io/cert-manager-webhook created
-------------------------------------stderr-------------------------------------------
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=<redacted> to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=<redacted> to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=<redacted> to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=<redacted> to perform this operation.
the namespace from the provided object "kube-system" does not match the namespace "ratingsapp". You must pass '--namespace=<redacted> to perform this operation.
the namespace from the provided object "kube-system" does not match the namespace "ratingsapp". You must pass '--namespace=<redacted> to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=<redacted> to perform this operation.
the namespace from the provided object "kube-system" does not match the namespace "ratingsapp". You must pass '--namespace=<redacted> to perform this operation.
the namespace from the provided object "kube-system" does not match the namespace "ratingsapp". You must pass '--namespace=<redacted> to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=<redacted> to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=<redacted> to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=<redacted> to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=<redacted> to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=<redacted> to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=<redacted> to perform this operation.
  (x) Failed: Deploying service api

ERROR: failed deploying service 'api': failing invoking action 'deploy', failed applying kube manifests: failed process templates, failed applying file 'src/manifests/api/1-cert-manager.yaml', kubectl apply -f: exit code: 1, stdout: namespace/cert-manager created
customresourcedefinition.apiextensions.k8s.io/certificaterequests.cert-manager.io created
customresourcedefinition.apiextensions.k8s.io/certificates.cert-manager.io created
customresourcedefinition.apiextensions.k8s.io/challenges.acme.cert-manager.io created
customresourcedefinition.apiextensions.k8s.io/clusterissuers.cert-manager.io created
customresourcedefinition.apiextensions.k8s.io/issuers.cert-manager.io created
customresourcedefinition.apiextensions.k8s.io/orders.acme.cert-manager.io created
clusterrole.rbac.authorization.k8s.io/cert-manager-cainjector created
clusterrole.rbac.authorization.k8s.io/cert-manager-controller-issuers created
clusterrole.rbac.authorization.k8s.io/cert-manager-controller-clusterissuers created
clusterrole.rbac.authorization.k8s.io/cert-manager-controller-certificates created
clusterrole.rbac.authorization.k8s.io/cert-manager-controller-orders created
clusterrole.rbac.authorization.k8s.io/cert-manager-controller-challenges created
clusterrole.rbac.authorization.k8s.io/cert-manager-controller-ingress-shim created
clusterrole.rbac.authorization.k8s.io/cert-manager-view created
clusterrole.rbac.authorization.k8s.io/cert-manager-edit created
clusterrole.rbac.authorization.k8s.io/cert-manager-controller-approve:cert-manager-io created
clusterrole.rbac.authorization.k8s.io/cert-manager-controller-certificatesigningrequests created
clusterrole.rbac.authorization.k8s.io/cert-manager-webhook:subjectaccessreviews created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-cainjector created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-controller-issuers created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-controller-clusterissuers created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-controller-certificates created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-controller-orders created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-controller-challenges created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-controller-ingress-shim created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-controller-approve:cert-manager-io created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-controller-certificatesigningrequests created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-webhook:subjectaccessreviews created
mutatingwebhookconfiguration.admissionregistration.k8s.io/cert-manager-webhook created
validatingwebhookconfiguration.admissionregistration.k8s.io/cert-manager-webhook created
, stderr: the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=cert-manager' to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=cert-manager' to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=cert-manager' to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=cert-manager' to perform this operation.
the namespace from the provided object "kube-system" does not match the namespace "ratingsapp". You must pass '--namespace=kube-system' to perform this operation.
the namespace from the provided object "kube-system" does not match the namespace "ratingsapp". You must pass '--namespace=kube-system' to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=cert-manager' to perform this operation.
the namespace from the provided object "kube-system" does not match the namespace "ratingsapp". You must pass '--namespace=kube-system' to perform this operation.
the namespace from the provided object "kube-system" does not match the namespace "ratingsapp". You must pass '--namespace=kube-system' to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=cert-manager' to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=cert-manager' to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=cert-manager' to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=cert-manager' to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=cert-manager' to perform this operation.
the namespace from the provided object "cert-manager" does not match the namespace "ratingsapp". You must pass '--namespace=cert-manager' to perform this operation.

WARNING: your version of azd is out of date, you have 1.3.0 and the latest version is 1.3.1

To update to the latest version, run:
curl -fsSL https://aka.ms/install-azd.sh | bash

If the install script was run with custom parameters, ensure that the same parameters are used for the upgrade. For advanced install instructions, see: https://aka.ms/azd/upgrade/linux

**Comment by cquinn-dcsg** [comment]
Running into bicep issues that work on other versions of azd:

cquinn@DESKTOP-72A3U61:.../repos$ azd version
azd version 1.4.0-beta.1-pr.3092531 (commit 74271d9013645c8cf9d0025d100b3a9f8a278b9d)


Provisioning Azure resources (azd provision)
Provisioning Azure resources can take some time

  (✓) Done: Downloading Bicep
? Select an Azure Subscription to use:  6. AKS-CSET-DEV (82e70289-bf40-45f9-8476-eab93d2031f4)
  (✓) Done: Retrieving locations...
? Select an Azure location to use: 41. (US) Central US (centralus)
  (✓) Done: Retrieving locations...

ERROR: deployment failed: failing invoking action 'provision', error deploying infrastructure: creating template: failed to compile bicep template: failed running bicep build: exit code: 1, stdout: , stderr: /mnt/c/Users/carmodyquinn/repos/ratings-nodejs-azd-akslza/infra/core/bicep/keyvault.bicep(58,57) : Warning use-recent-api-versions: Use more recent API version for 'Microsoft.Insights/diagnosticSettings'. '2021-05-01-preview' is 877 days old, should be no more than 730 days old, or the most recent. Acceptable versions: 2016-09-01 [https://aka.ms/bicep/linter/use-recent-api-versions]

**Comment by carmodyquinnms** [comment]
@wbreza - I found that there was a bicep issue with the latest bicep.  I fixed that in my repo.  The PR is tested and ready for a merge (with the caveat that at the end of the provision step it hangs waiting for login, but I believe you know that and that it will be fixed with the azd aad updates).  So this can be merged and once merged I will promote my branch to main.

File: cli/azd/pkg/tools/kubectl/kubectl.go
Comment (weikanglim) [inline cli/azd/pkg/tools/kubectl/kubectl.go:341]: This does end up modifying the `cli.env` map. We could make `SetKubeConfig` modify `env` directly instead to express the same idea more cleanly.

I do feel long-term there should be a more uniform way for us to express the idea in code:
 - For given tool X, environment variables `ENV_FOO`, `ENV_BAR` will be passed through in azd, that doesn't require the more elaborate setup.
I wonder if we feel this is something worth exploring.

Comment (wbreza) [reply cli/azd/pkg/tools/kubectl/kubectl.go:341]: I like this strategy and updated `SetKubeConfig` to just modify the `cli.env` directly for now.  Will hold off on a larger strategy for a future PR.

**Review by weikanglim — APPROVED** [review]

**Pull request merged**
